flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeStore.java

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * A {@link StateChangeStore} that waits for some configured amount of time before passing the
+ * accumulated state changes to the actual store.
+ */
+class BatchingStateChangeStore implements StateChangeStore {
+    private static final Logger LOG = LoggerFactory.getLogger(BatchingStateChangeStore.class);
+
+    private final ScheduledExecutorService scheduler;
+    private final long scheduleDelayMs;
+    private final BlockingQueue<StateChangeSet> scheduled;
+    private final AtomicBoolean drainScheduled;
+    private final RetryPolicy retryPolicy;
+    private volatile Throwable error;
+    private final int sizeThreshold;
+    private final RetryingExecutor retryingExecutor;
+    private final StateChangeStore delegate;
+
+    BatchingStateChangeStore(
+            long persistDelayMs,
+            int sizeThreshold,
+            int requestQueueCapacity,
+            RetryPolicy retryPolicy,
+            StateChangeStore delegate) {
+        this.scheduleDelayMs = persistDelayMs;
+        this.scheduled = new ArrayBlockingQueue<>(requestQueueCapacity, true);
+        this.scheduler = SchedulerFactory.create(1, "ChangelogRetryScheduler", LOG);
+        this.drainScheduled = new AtomicBoolean(false);
+        this.retryPolicy = retryPolicy;
+        this.retryingExecutor = new RetryingExecutor();
+        this.sizeThreshold = sizeThreshold;
+        this.delegate = delegate;
+    }
+
+    BatchingStateChangeStore(
+            long persistDelayMs,
+            int sizeThreshold,
+            int requestQueueCapacity,
+            RetryPolicy retryPolicy,
+            StateChangeStore delegate,
+            ScheduledExecutorService scheduler,
+            RetryingExecutor retryingExecutor) {
+        this.scheduleDelayMs = persistDelayMs;
+        this.scheduled = new ArrayBlockingQueue<>(requestQueueCapacity, true);
+        this.scheduler = scheduler;
+        this.drainScheduled = new AtomicBoolean(false);
+        this.retryPolicy = retryPolicy;
+        this.retryingExecutor = retryingExecutor;
+        this.sizeThreshold = sizeThreshold;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void save(Collection<StateChangeSet> changeSets) {
+        if (error != null) {
+            LOG.debug("don't persist {} changesets, already failed", changeSets.size());
+            changeSets.forEach(cs -> cs.setFailed(error));
+            return;
+        }
+        LOG.debug("persist {} changeSets", changeSets.size());
+        try {
+            for (StateChangeSet changeSet : changeSets) {
+                scheduled.put(changeSet); // blocks if no space in the queue
+            }
+            scheduleUploadIfNeeded();
+        } catch (Exception e) {
+            changeSets.forEach(cs -> cs.setFailed(e));
+            rethrow(e);
+        }
+    }
+
+    private void scheduleUploadIfNeeded() {
+        if (scheduleDelayMs == 0 || scheduled.size() >= sizeThreshold) {
```

Comment:
I agree that size in bytes is preferrable. But it would introduce complexity because it wouldn't be atomic and would be accessed from different threads. 
For MVP, timeout plus queueSizeThreshold should be enough IMO (I still would like this threshold to be configurable).